contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,167 @@
+//SPDX-License-Identifier: MIT
+pragma solidity 0.8.18;
+
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "@subnet-evm-contracts/interfaces/IWarpMessenger.sol";
+import "@subnet-evm-contracts/interfaces/IAllowList.sol";
+import "@subnet-evm-contracts/interfaces/INativeMinter.sol";
+import "./INativeTokenDestination.sol";
+import "../../Teleporter/ITeleporterMessenger.sol";
+import "../../Teleporter/ITeleporterReceiver.sol";
+import "../../Teleporter/SafeERC20TransferFrom.sol";
+
+// Precompiled Native Minter Contract Address
+address constant MINTER_ADDRESS = 0x0200000000000000000000000000000000000001;
+// Designated Blackhole Address
+address constant BLACKHOLE_ADDRESS = 0x0100000000000000000000000000000000000000;
+
+contract NativeTokenDestination is
+    ITeleporterReceiver,
+    INativeTokenDestination,
+    ReentrancyGuard
+{
+    INativeMinter private immutable _nativeMinter =
+        INativeMinter(MINTER_ADDRESS);
+
+    address public constant WARP_PRECOMPILE_ADDRESS =
+        0x0200000000000000000000000000000000000005;
+
+    uint256 public constant TRANSFER_NATIVE_TOKENS_REQUIRED_GAS = 10_000_000; // TODO this is a placeholder
+    bytes32 public immutable currentChainID;
+    bytes32 public immutable sourceChainID;
+    address public immutable sourceContractAddress;
+
+    // Used for sending an receiving Teleporter messages.
+    ITeleporterMessenger public immutable teleporterMessenger;
+
+    error InvalidTeleporterMessengerAddress();
+    error InvalidRecipientAddress();
+    error InvalidSourceChain();
+    error InvalidPartnerContractAddress();
+    error CannotBridgeTokenWithinSameChain();
+    error Unauthorized();
+    error InsufficientPayment();
+    error InvalidSourceContractAddress();
+    error InsufficientAdjustedAmount(uint256 adjustedAmount, uint256 feeAmount);
+
+    // TODO we probably want to add the original token supply from this chain to the constructor.
+    constructor(
+        address teleporterMessengerAddress,
+        address sourceContractAddress_,
+        bytes32 sourceChainID_
```

Comment:
Nit: For consistent ordering, lets but the chain ID prior to the contract address.